### DISCUSSION TITLE
How to pass control to useController in React Typescript?

### DISCUSSION THREAD

**Discussion by hendra-foo** [react-hook-form/react-hook-form#5983] [Q&A] [answered]
Hello, I'm trying to make a reusable input type text with `useController` in React Typescript

when I try to pass the `control` returned from `useForm` there is an error saying:
`Type 'Control<FormValues>' is not assignable to type 'Control<Record<string, any>>'.`
 `Types of property 'register' are incompatible.`
  ` Type 'UseFormRegister<FormValues>' is not assignable to type 'UseFormRegister<Record<string, any>>'.`
   ` Type 'Record<string, any>' is not assignable to type 'FormValues'.ts(2322)`

am I doing wrong with my `FormValues` Type?
sandbox link: https://codesandbox.io/s/control-forked-z65ek?file=/src/App.tsx

**Comment by michalbundyra** [comment]
The same issue here. I noticed it works when we add `Partial` for the form data type:

```typescript
useForm<Partial<FormValues>>(...)
```

but this is not perfect.

**Reply by hendra-foo** [reply]
yes, I'm currently using the same solutions as you by passing `Partial<Type>` to the `useForm` generic params,

but this will lead to another issue when I try to use the `useFieldArray` hooks
the returned `fields` type is incorrect, it says `never[]`

`useFieldArray`
docs: https://react-hook-form.com/api/usefieldarray
example: https://codesandbox.io/s/react-hook-form-usefieldarray-forked-14c8d?file=/src/index.js

updated sandbox link:
https://codesandbox.io/s/control-forked-e43wk?file=/src/App.tsx

**Comment by hendra-foo** [comment]
just forked `useController` demo from the docs and migrated to typescript version.
the form is functionally properly, but the type error is still there
https://codesandbox.io/s/usecontroller-forked-dvmr8?file=/src/App.tsx

I feel so bad to have to bother you here @bluebill1049 , but can you please help check the sandbox when you have some free time?
I haven't found any better solutions to this yet. Thank you!

**Reply by bluebill1049** [reply]
hey @hendra-foo sorry I completely miss your comment (too many notifications)

```tsx
import * as React from "react";
import { useForm, useController, UseControllerProps } from "react-hook-form";
import "./styles.css";

type FormValues = {
  FirstName: string
}

function Input(props: UseControllerProps<FormValues>) {
  const { field, fieldState } = useController(props);

  return (
    <div>
      <input {...field} placeholder={props.name} />
      <p>{fieldState.isTouched && "Touched"}</p>
      <p>{fieldState.isDirty && "Dirty"}</p>
      <p>{fieldState.invalid ? "invalid" : "valid"}</p>
    </div>
  );
}

export default function App() {
  const { handleSubmit, control, watch } = useForm<FormValues>({
    defaultValues: {
      FirstName: ""
    },
    mode: "onChange"
  });
  const onSubmit = (data) => console.log(data);

  console.log(watch());

  return (
    <div>
      <h1>useController</h1>
      <form onSubmit={handleSubmit(onSubmit)}>
        <Input control={control} name="FirstName" rules={{ required: true }} />
        <input type="submit" />
      </form>
    </div>
  );
}
```

try the above.


and this should work as well

```tsx
import * as React from "react";
import { useForm, useController, UseControllerProps } from "react-hook-form";
import "./styles.css";

type FormValues = {
  FirstName: string
}

function Input<T>(props: UseControllerProps<T>) {
  const { field, fieldState } = useController(props);

  return (
    <div>
      <input {...field} placeholder={props.name as string} />
      <p>{fieldState.isTouched && "Touched"}</p>
      <p>{fieldState.isDirty && "Dirty"}</p>
      <p>{fieldState.invalid ? "invalid" : "valid"}</p>
    </div>
  );
}

export default function App() {
  const { handleSubmit, control, watch } = useForm<FormValues>({
    defaultValues: {
      FirstName: ""
    },
    mode: "onChange"
  });
  const onSubmit = (data) => console.log(data);

  console.log(watch());

  return (
    <div>
      <h1>useController</h1>
      <form onSubmit={handleSubmit(onSubmit)}>
        <Input<FormValues> control={control} name="FirstName" rules={{ required: true }} />
        <input type="submit" />
      </form>
    </div>
  );
}
```

hope they are helpful. 🙏

**Comment by hendra-foo** [comment, marked as answer]
@michalbundyra 
fyi, incase you haven't solve this issue yet,
pass your `control` type to the `useControllerProps` generic params

for further details you can look at this existing discussions:
https://github.com/react-hook-form/react-hook-form/discussions/3510#discussioncomment-137613
